### ISSUE TITLE
[API Proposal]: ImmutableInterlocked.VolatileIsDefault

### ISSUE THREAD

**Issue by 333fred** [dotnet/runtime#109572]
### Background and motivation

In Roslyn, we love our lockless algorithms; we do not mind occasionally doing extra work and throwing away the extra work in order to avoid locks. We usually accomplish this via sentinel fields; we initialize the field via some `Interlocked` API (usually `CompareExchange` or `Initialize`), and then read that field later using a `Volatile.Read`. Any other accesses to other fields are dependent on that read: we only treat other fields as valid if that read returns valid info. Unfortunately, we want to (and sadly do) use `ImmutableArray<T>`s as sentinels, but they don't have a method of safely doing this `Volatile.Read`. This leads to [theoretical race conditions](https://github.com/dotnet/roslyn/issues/75759) that, while incredibly unlikely, do exist on some architectures. We'd like a way to correctly have a read fence for these scenarios. I'm suggesting one of two possibilities to start with; I don't have strong opinions on which direction is best for the API overall. Either will work.

#### Option 1

```csharp
namespace System.Collections.Immutable;

public static class ImmutableInterlocked
{
    public bool VolatileIsDefault<T>(ref ImmutableArray<T> location) => Volatile.Read(ref Unsafe.AsRef(in location.array)).IsDefault;
}
```

#### Option 2

```csharp
namespace namespace System.Runtime.InteropServices;

public static class ImmutableCollectionsMarshal
{
    public ref T[] AsArrayRef<T>(ref ImmutableArray<T> location) => ref Unsafe.AsRef(in array.array);
}
```

#### Option 1

```cs
if (ImmutableInterlocked.VolatileIsDefault(ref _lazyDeclaredConstraintTypes))
{
    // Init code including other fields
    ImmutableInterlocked.InterlockedInitialize(ref _lazyDeclaredConstraintTypes, declaredConstraintTypes);

}

// Access other fields knowing they're guaranteed to be valid at this point
```

#### Option 2

```cs
if (Volatile.Read(ref ImmutableCollectionsMarshal.AsRef(ref _lazyDeclaredConstraintTypes)).IsDefault)
{
    // Init code including other fields
    ImmutableInterlocked.InterlockedInitialize(ref _lazyDeclaredConstraintTypes, declaredConstraintTypes);

}

// Access other fields knowing they're guaranteed to be valid at this point
```

**Comment by EgorBo** [comment]
The original issue mentions `Interlocked.MemoryBarrier()` and `Technically has some runtime impact, but not noticeable.` if that is the case, is a new API even needed?

**Comment by 333fred** [comment]
While it does work, it doesn't feel as clean, and moves the safety farther (syntax-wise) from the thing that relies on it. I'd definitely prefer to have the guard on the actual operation, rather than a distance away. I will also say that I don't think either option here exposes new internal data structures to the world. Conceptually, `VolatileIsDefault` doesn't specify how it actually does the check, only that it introduces a read barrier. And we already have `ImmutableCollectionsMarshal.AsArray` and `ImmutableCollectionsMarshal.AsImmutableArray`, so we already have the "the backing item is an array" is already pretty well exposed.

**Comment by timcassell** [comment]
`Volatile.ReadBarrier` was recently added (#107843). I imagine a un-constrained `Volatile.Read<T>` (#107449) could special-case such simple wrappers, and use the new `ReadBarrier` API for larger structs.

**Comment by EgorBo** [comment]
> And we already have ImmutableCollectionsMarshal.AsArray and ImmutableCollectionsMarshal.AsImmutableArray, so we already have the "the backing item is an array" is already pretty well exposed.

Ah I see. It just feels like we can open a can of worms if we start adding APIs allowing to access internal fields with extra barriers. I can see how other users use different data structure/classes as sentinels.

Btw, is an extra `bool isInitialized` field a bad idea for your cases?

**Comment by jkotas** [comment]
> Conceptually, VolatileIsDefault doesn't specify how it actually does the check, only that it introduces a read barrier.

Right, it is two separate operation: IsDefault check and a read barrier. We have introduced `Interlocked.ReadBarrier` to allow you to combine an arbitrary operation with read barrier in any way you wish. We can introduce a convenience APIs for popular combinations of an operation and a barrier, but that's just nice to have. `VolatileIsDefault` can be implemented like this using the `Interlocked.ReadBarrier` API. Notice that the location argument does not even need to be passed by `ref` for it to work:

```
public bool VolatileIsDefault<T>(ImmutableArray<T> location)
{
  bool ret = location.IsDefault;
  Interlocked.ReadBarrier();
  return ret;
}
```

**Comment by 333fred** [comment]
My one grumble about `Volatile.ReadBarrier()` is that I can't use it on all of Roslyn's TFMs, but I don't think that this is important enough to add if the BCL's position would otherwise be "write the extension method yourself using `Volatile.ReadBarrier()`, we won't add it". We can just use `Interlocked.MemoryBarrier()` on downlevel platforms that don't have `Volatile.ReadBarrier`. I _would_ still like this if y'all are willing to add it, but it's not the end of the world.

**Comment by 333fred** [comment]
I've added a 3rd API option that came up during a compiler team meeting on this topic, suggested by @AlekseyTs. We're planning on adding a version of this API to our own `RoslynImmutableInterlocked` type, but would really like to see this added to `ImmutableInterlocked`. The BCL currently ships built-in Immutable APIs for part of a lockless design (read/write barriers for initialization), but not corresponding APIs for reading that array safely.
